### PR TITLE
K8SPSMDB-578-579: check both SSL secret and SSL internal secret in reconsileSSL

### PR DESCRIPTION
**Pull request by srteam2020**
[K8SPSMDB-578](https://jira.percona.com/browse/K8SPSMDB-578) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=percona&utm_campaign=badge_info)

- Fixed K8SPSMDB-578
- Fixed K8SPSMDB-579

There is bug that prevents `ssl-internal` secret from getting created if operator crashs in the middle of reconcile.
Simpilfied reconcile logic is shown below,
```go
secretObj := corev1.Secret{}
err := r.client.Get(context.TODO(), ...) // Get ssl secret
if(notFoundSecret){
    r.client.Create(secretObj)  // (a)
    r.client.Create(secretInteralObj)  // (b)
}
```
Suppose if the operator crash after (a) but before (b). In that case, after the operator gets recovered the `secretObj` can be found, and the `secretInteralObj` will never be created.

This patch changes it to:
```go
secretObj := corev1.Secret{}
err := r.client.Get(context.TODO(), ...) // Get ssl secret
secretInternalObj := corev1.Secret{}
err := r.client.Get(context.TODO(), ...) // Get ssl internal secret
if(notFoundSecret || notFoundInternalSecret){
    r.client.Create(secretObj)  // (a)
    r.client.Create(secretInteralObj)  // (b)
}
```
which can fix the bug.

### CODE REVIEW

File: pkg/controller/perconaservermongodb/ssl.go [hunk 1]
```diff
@@ -20,19 +20,29 @@ import (
 
 func (r *ReconcilePerconaServerMongoDB) reconsileSSL(cr *api.PerconaServerMongoDB) error {
 	secretObj := corev1.Secret{}
-	err := r.client.Get(context.TODO(),
+	secretInternalObj := corev1.Secret{}
```
Comment (hors) [inline pkg/controller/perconaservermongodb/ssl.go:23]: @defbin please add good description to your PR as it was decided

**Review by hors — CHANGES_REQUESTED** [review]

File: pkg/controller/perconaservermongodb/ssl.go (hunk 1 above)
Comment (srteam2020) [reply pkg/controller/perconaservermongodb/ssl.go:23]: Sure, I have updated it in the description.

Comment (srteam2020) [reply pkg/controller/perconaservermongodb/ssl.go:23]: Sure, I have updated it in the description.

Comment (srteam2020) [reply pkg/controller/perconaservermongodb/ssl.go:23]: Yes I just added a description

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | failed |
| init-deploy | passed |
| storage | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| limits | passed |
| demand-backup | failed |
| scaling | passed |
| security-context | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| rs-shard-migration | passed |
| monitoring-2-0 | passed |
| scheduled-backup | failed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| demand-backup-sharded | failed |
| users | failed |
| data-sharded | passed |
| upgrade | failed |
| non-voting | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/797/commits/39dee19b75b1572dd9e966a52eec75316c2bbdc9
image: `perconalab/percona-server-mongodb-operator:PR-797-39dee19b`

**Comment by srteam2020** [comment]
We are not able to see the CI test failure details and are unsure whether it is caused by the change in the PR

**Comment by egegunes** [comment]
> We are not able to see the CI test failure details and are unsure whether it is caused by the change in the PR

@srteam2020 The failures seem to be caused by these changes. Cluster is failing to start. Could you please run the failing tests on your environment?

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade-consistency | failed |
| self-healing | passed |
| self-healing-chaos | failed |
| limits | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| demand-backup-sharded | passed |
| smart-update | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | failed |
| version-service | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| users | passed |
| pitr | failed |
| data-sharded | passed |
| non-voting | passed |
| pitr-sharded | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/797/commits/6a76ded17a65ef55945b8520030eef26eb236381
image: `perconalab/percona-server-mongodb-operator:PR-797-6a76ded1`

**Comment by srteam2020** [comment]
Hello how is the status of this PR?

**Comment by egegunes** [comment]
> Hello how is the status of this PR?

Hello @srteam2020, the fix was waiting the tests to pass. Did you have the chance to run tests on your environment? 

Having said that, we discussed the issue with team. If SSL secrets exists, not creating SSL internal secret is intentional. The operator will use SSL secrets everywhere in this case. It seems it's not a bug.

**Comment by srteam2020** [comment]
Hello @egegunes Thanks for the reply! Yes we can try to run the tests on our environment.

Similar as discussed in another thread,
we found that Mongodb can continue running even without the SSL internal certificate but this behavior actually cause problems to the users -- We got confused by this problem and tried to debug this for days. Here are the reasons why we think this can actually cause problem and the checking logic should be improved:
1. As hinted both by the code (`createSSLManualy` always creates both secrets) and the docuementatin (https://www.percona.com/doc/kubernetes-operator-for-psmongodb/TLS.html), only creating one secret is an undesired behavior to us (unless this is explicitly specified by the user). We believe the current `createSSLManualy` is doing correctly by creating both secrets and what should be improved is the checking logic in `reconsileSSL`.
2. Using different certificates (secrets) for external/internal communications provides very strong security gaurantee and we really appreciate this feature. However, the imcomplete checking logic in `reconsileSSL` makes the operator unable to create the internal one which unnecessarily sacrifices the security, and this issue can be totally prevented by fixing the checking logic.

We enjoyed using the mongodb operator and really appreciate your awesome work on this! So we really hope we could help on improving the reliability and security. This issue indeed causes much confusion and debugging work for us so we would really appreciate if the checking logic can be improved and two secretes are always created. So that no one will encounter the same issue in the future.

Please let us know your further concerns on the fix. We are willing to improve the PR.

**Comment by egegunes** [comment]
@srteam2020 I'll discuss it with the team

**Comment by egegunes** [comment]
@srteam2020, we believe your points are valid but the fix needs to support two scenarios:

1. User creates the TLS secret manually and passes its name to CR
2. Operator creates the TLS secrets automatically

Your fix is OK for the scenario 2 but on scenario 1 operator needs to use the same secret for both internal and external communications. If you could check the owner references to detect if secret created by the user or the operator, we can cover these two scenarios.

**Comment by srteam2020** [comment]
@egegunes Thanks for the reply!
If I understand correctly, the operator should check the owner of the first secret (the external one) to see whether it is created by the user or by the operator. The operator only needs to create the second secrete (the internal one) if the first one is automatically created by the operator.
I can go ahead to update the PR if my understanding is correct.

**Comment by egegunes** [comment]
@srteam2020 yes, it's correct. please let me know if you have any questions

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| limits | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| scheduled-backup | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| liveness | passed |
| demand-backup-sharded | passed |
| smart-update | passed |
| version-service | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| users | passed |
| pitr | passed |
| data-sharded | passed |
| non-voting | passed |
| pitr-sharded | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/797/commits/2924ca9991ef84abb9f9ac581163652ef31605fa
image: `perconalab/percona-server-mongodb-operator:PR-797-2924ca99`

File: pkg/controller/perconaservermongodb/ssl.go [hunk 2]
```diff
@@ -20,19 +20,46 @@ import (
...
+		}
+		for _, owner := range secretObj.OwnerReferences {
+			if owner.UID == ownerCr.UID {
+				ownedByCr = true
+				break
+			}
+		}
+		if !ownedByCr {
+			return nil
+		}
```
Comment (egegunes) [inline pkg/controller/perconaservermongodb/ssl.go:60]: it's better to use https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#IsControlledBy

**Review by egegunes — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| limits | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| scheduled-backup | passed |
| rs-shard-migration | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| data-sharded | passed |
| pitr | passed |
| non-voting | passed |
| pitr-sharded | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/797/commits/ac2177f1f8d4270cf91c3b8b4bc30ddaa7823ee2
image: `perconalab/percona-server-mongodb-operator:PR-797-ac2177f1`

File: pkg/controller/perconaservermongodb/ssl.go (hunk 2 above)
Comment (srteam2020) [reply pkg/controller/perconaservermongodb/ssl.go]: Fixed

**Comment by srteam2020** [comment]
I am not sure whether the failed `pitr-sharded` is due to the PR. Should we rebase main and rerun the tests?

**Review by egegunes — APPROVED** [review]
@srteam2020 `pitr-sharded` is a bit flaky due to timeouts. I re-ran the tests

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| init-deploy | passed |
| one-pod | passed |
| limits | passed |
| monitoring-2-0 | passed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| service-per-pod | passed |
| rs-shard-migration | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| non-voting | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | passed |
| pitr-sharded | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/797/commits/ac2177f1f8d4270cf91c3b8b4bc30ddaa7823ee2
image: `perconalab/percona-server-mongodb-operator:PR-797-ac2177f1`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| limits | passed |
| scaling | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | failed |
| monitoring-2-0 | passed |
| security-context | passed |
| rs-shard-migration | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| arbiter | passed |
| upgrade-sharded | passed |
| service-per-pod | passed |
| pitr | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| pitr-sharded | failed |
| users | passed |
| data-sharded | failed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/797/commits/2570d84ebe69a74710d226a19ab71e7f5dec37c9
image: `perconalab/percona-server-mongodb-operator:PR-797-2570d84e`

**Comment by srteam2020** [comment]
@egegunes Are these failed tests also flaky? Or is there anything else I should do to help merge this PR?

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade-consistency | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| limits | passed |
| monitoring-2-0 | failed |
| demand-backup | passed |
| scaling | passed |
| security-context | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| scheduled-backup | passed |
| rs-shard-migration | passed |
| arbiter | passed |
| demand-backup-sharded | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| upgrade | passed |
| version-service | passed |
| upgrade-sharded | passed |
| pitr | passed |
| users | passed |
| data-sharded | passed |
| pitr-sharded | passed |
| non-voting | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/797/commits/0f737c615c0548638038b31b4cc474419f865993
image: `perconalab/percona-server-mongodb-operator:PR-797-0f737c61`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| limits | passed |
| monitoring-2-0 | failed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| storage | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | passed |
| pitr-sharded | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| non-voting | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/797/commits/0f737c615c0548638038b31b4cc474419f865993
image: `perconalab/percona-server-mongodb-operator:PR-797-0f737c61`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| limits | passed |
| monitoring-2-0 | failed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| storage | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | passed |
| pitr-sharded | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| non-voting | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/797/commits/0f737c615c0548638038b31b4cc474419f865993
image: `perconalab/percona-server-mongodb-operator:PR-797-0f737c61`

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| monitoring-2-0 | failed |
| limits | passed |
| scaling | passed |
| security-context | passed |
| rs-shard-migration | passed |
| storage | passed |
| self-healing | passed |
| self-healing-chaos | passed |
| operator-self-healing | passed |
| operator-self-healing-chaos | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |
| upgrade | passed |
| upgrade-sharded | passed |
| pitr | passed |
| pitr-sharded | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |
| data-sharded | passed |
| non-voting | passed |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/797/commits/0f737c615c0548638038b31b4cc474419f865993
image: `perconalab/percona-server-mongodb-operator:PR-797-0f737c61`

**Comment by JNKPercona** [comment]
| Test name | Status |
| ------------- | ------------- |
| init-deploy | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/init-deploy.log) |
| one-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/one-pod.log) |
| limits | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/limits.log) |
| monitoring-2-0 | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/monitoring-2-0.log) |
| scaling | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/scaling.log) |
| arbiter | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/arbiter.log) |
| storage | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/storage.log) |
| security-context | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/security-context.log) |
| service-per-pod | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/service-per-pod.log) |
| self-healing | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/self-healing.log) |
| rs-shard-migration | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/rs-shard-migration.log) |
| liveness | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/liveness.log) |
| self-healing-chaos | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/self-healing-chaos.log) |
| smart-update | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/smart-update.log) |
| operator-self-healing | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/operator-self-healing.log) |
| version-service | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/version-service.log) |
| operator-self-healing-chaos | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/operator-self-healing-chaos.log) |
| users | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/users.log) |
| data-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/data-sharded.log) |
| non-voting | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/non-voting.log) |
| upgrade-consistency | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/upgrade-consistency.log) |
| demand-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/demand-backup.log) |
| scheduled-backup | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/scheduled-backup.log) |
| demand-backup-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/demand-backup-sharded.log) |
| upgrade | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/upgrade.log) |
| upgrade-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/upgrade-sharded.log) |
| pitr | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/pitr.log) |
| pitr-sharded | [passed](https://percona-jenkins-artifactory-public.s3.amazonaws.com/cloud-psmdb-operator/PR-797/7f6764ba/pitr-sharded.log) |

commit: https://github.com/percona/percona-server-mongodb-operator/pull/797/commits/7f6764ba856b21151d35bbaabacb58d835c73112
image: `perconalab/percona-server-mongodb-operator:PR-797-7f6764ba`

**Review by hors — APPROVED** [review]

**Pull request merged**